File: 78-Subsets.py
class Solution(object):
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
# DFS O(N*2^N)
        res = []
        n = len(nums)
        def dfs(val,start):
            res.append(val)
            for i in range(start,n):
                dfs(val+[nums[i]], i+1)
        dfs([],0)
        return res
            

File: C++/78-Subsets.cpp
class Solution {
    vector<vector<int>> res;
    int N;
public:
    void dfs(int idx, vector<int>& arr, vector<int>& nums) {
        if (idx == N) {
            res.push_back(arr);
            return;
        }
        dfs(idx+1, arr, nums);
        arr.push_back(nums[idx]);
        dfs(idx+1, arr, nums);
        arr.pop_back();
    }
    vector<vector<int>> subsets(vector<int>& nums) {
        N = nums.size();
        vector<int> arr;
        dfs(0, arr, nums);
        return res;
    }
};